### PR TITLE
feat: Java function support for CLI

### PR DESCRIPTION
**Pull request by akshbhu**
*Description of changes:*
* added java building and packaging
* added helloworld template 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-function-plugin-interface/src/index.ts [hunk 1]
```diff
@@ -80,6 +80,7 @@ export type BuildResult = {
 // Result of packaging a function
 export type PackageResult = {
   packageHash?: string; // undefined if no repacking necessary. Otherwise, it is a hash that uniquiely identifies the package
+  zipFileName?: string; // defined for certain plugins like Java
```
Comment (edwardfoyle) [inline packages/amplify-function-plugin-interface/src/index.ts]: The dstFilename in the PackageRequest is the zip file. Do you need a different one?

File: packages/amplify-java-runtime-provider/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+       return Promise.resolve({
+         runtime: {
+           name: 'Java',
+           value: 'java',
+           cloudTemplateValue: 'java8',
+           defaultHandler: 'example.HelloPojo',
+         },
+      });
+    },
+    checkDependencies: () => Promise.resolve({ hasRequiredDependencies: true }),
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/index.ts]: Probably want to actually make sure java and javac are on the path

File: packages/amplify-java-runtime-provider/src/utils/build.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  const resourceDir = path.join(request.srcRoot);
+
+  if (!request.lastBuildTimestamp || isValidProject(resourceDir)) {
+    installDependencies(resourceDir);
+    return Promise.resolve({ rebuilt: true });  }
+    return Promise.resolve({ rebuilt: false });
+  }
+
+function isValidProject(cwd : string){
+  const packageManager = 'mvn';
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/build.ts]: I thought we decided on supporting gradle builds?

File: packages/amplify-java-runtime-provider/src/utils/build.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import path from 'path';
+import fs from 'fs-extra';
+import childProcess from 'child_process';
+import { BuildRequest, BuildResult } from 'amplify-function-plugin-interface';
+
+// copied from the existing build-resources.js file in amplify-cli with changes for new interface
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/build.ts]: remove comment

File: packages/amplify-java-runtime-provider/src/utils/build.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+export async function buildResource(request: BuildRequest): Promise<BuildResult>{
+  const resourceDir = path.join(request.srcRoot);
+
+  if (!request.lastBuildTimestamp || isValidProject(resourceDir)) {
+    installDependencies(resourceDir);
+    return Promise.resolve({ rebuilt: true });  }
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/build.ts]: formatting. Also this function should return rebuilt: true iff it did a fresh install / compile. Otherwise return rebuilt: false

File: packages/amplify-java-runtime-provider/src/utils/package.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import childProcess from 'child_process';
+import path from 'path';
+import { PackageRequest, PackageResult } from 'amplify-function-plugin-interface/src';
+
+export async function packageResource(request: PackageRequest, context: any): Promise<PackageResult> {
+  const resourceDir = path.join(request.srcRoot);
+  const meta = context.amplify.getProjectDetails().amplifyMeta;
+  const zipFileName = Object.keys(meta.function)[0] + '-1.0-SNAPSHOT.jar';
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/package.ts]: you should be zipping into request.dstFilename. Otherwise the platform won't know where to find the zip file

File: packages/amplify-java-runtime-provider/src/utils/build.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import fs from 'fs-extra';
+import childProcess from 'child_process';
+import { BuildRequest, BuildResult } from 'amplify-function-plugin-interface';
+
+// copied from the existing build-resources.js file in amplify-cli with changes for new interface
+export async function buildResource(request: BuildRequest): Promise<BuildResult>{
+  const resourceDir = path.join(request.srcRoot);
+
+  if (!request.lastBuildTimestamp || isValidProject(resourceDir)) {
+    installDependencies(resourceDir);
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/build.ts]: should there also be a compilation step here?

File: packages/amplify-java-runtime-provider/src/utils/build.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import path from 'path';
+import fs from 'fs-extra';
+import childProcess from 'child_process';
+import { BuildRequest, BuildResult } from 'amplify-function-plugin-interface';
+
+// copied from the existing build-resources.js file in amplify-cli with changes for new interface
+export async function buildResource(request: BuildRequest): Promise<BuildResult>{
+  const resourceDir = path.join(request.srcRoot);
+
+  if (!request.lastBuildTimestamp || isValidProject(resourceDir)) {
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/build.ts]: does isValidProject() check for changes to the function? We should only be rebuilding if we need to

File: packages/amplify-java-runtime-provider/src/utils/package.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import childProcess from 'child_process';
+import path from 'path';
+import { PackageRequest, PackageResult } from 'amplify-function-plugin-interface/src';
+
+export async function packageResource(request: PackageRequest, context: any): Promise<PackageResult> {
+  const resourceDir = path.join(request.srcRoot);
+  const meta = context.amplify.getProjectDetails().amplifyMeta;
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/package.ts]: If you need access to the function name, we should update the PackageRequest. However, it doesn't seem like building / packaging should depend on the function name?

File: packages/amplify-java-template-provider/amplify-plugin.json [hunk 10]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "functionTemplate",
+  "id": "d579f2fa-35c1-49c4-950f-b588e779b4dd",
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/amplify-plugin.json]: latest version of the plugin interface doesn't require this ID anymore

File: packages/amplify-java-template-provider/amplify-plugin.json [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      "service": "Lambda",
+      "runtime": "java"
+    },
+    "templates": [
+      {
+        "name": "Hello World",
+        "value": "helloworld"
+      },
+      {
+        "name": "CRUD function for DynamoDB (Integration with API Gateway)",
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/amplify-plugin.json]: remove any templates you haven't actually implemented

File: packages/amplify-java-template-provider/resources/lambda/hello-world/HelloPojo.java.ejs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+<%= props.topLevelComment %>
+
+package example;
+
+import com.amazonaws.services.lambda.runtime.Context; 
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+
+public class HelloPojo implements RequestHandler<RequestClass, ResponseClass>{
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/resources/lambda/hello-world/HelloPojo.java.ejs]: This isn't a POJO

File: packages/amplify-java-template-provider/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { FunctionTemplateContributorFactory } from 'amplify-function-plugin-interface';
+
+import { provideHelloWorld } from './providers/helloWorldProvider';
+
+export const functionTemplateContributorFactory: FunctionTemplateContributorFactory = context => {
+  return {
+    contribute: selection => {
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/src/index.ts]: rebase with latest change to master (the selection is now a request object with the selection string as a property)

File: packages/amplify-java-template-provider/src/utils/destFileMapper.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import _ from 'lodash';
+import path from 'path';
+
+// Converts files to a map of file to destination filename
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/src/utils/destFileMapper.ts]: update comment. Althought I'm not sure if you should be using this method. I made it to work around the fact that existing node functions nest everything in a 'src' folder but going forward we shouldn't do that.

File: packages/amplify-java-template-provider/resources/lambda/hello-world/HelloPojo.java.ejs (hunk 12 above)
Comment (akshbhu) [reply packages/amplify-java-template-provider/resources/lambda/hello-world/HelloPojo.java.ejs]: I know this is not POJO , but the underlying request and response classes , this is just an example to show POJO example

File: packages/amplify-function-plugin-interface/src/index.ts (hunk 1 above)
Comment (akshbhu) [reply packages/amplify-function-plugin-interface/src/index.ts]: nope, changed it

File: packages/amplify-java-runtime-provider/src/index.ts (hunk 2 above)
Comment (akshbhu) [reply packages/amplify-java-runtime-provider/src/index.ts]: done

File: packages/amplify-java-runtime-provider/src/utils/build.ts (hunk 3 above)
Comment (akshbhu) [reply packages/amplify-java-runtime-provider/src/utils/build.ts]: changed from maven to gradle builds

File: packages/amplify-java-runtime-provider/src/utils/build.ts (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-java-runtime-provider/src/utils/build.ts]: removed

File: packages/amplify-java-runtime-provider/src/utils/package.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+    return new Promise((resolve, reject) => {
+      output.on('close', () => {
+        resolve({ packageHash });
+      });
+      output.on('error', err => {
+        reject(new Error(`Failed to copy zip with error: [${err}]`));
+      });
+      // buld through gradle build
+      let zipFile : string = "latest_build.zip";
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/package.ts]: The file name should come from the package request

File: packages/amplify-java-runtime-provider/tsconfig.json [hunk 16]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "outDir": "./lib",
+    "rootDir": "./src",
+    "composite": true,
+    "strict": true,
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/tsconfig.json]: both of these values are set in the root tsconfig

File: packages/amplify-java-template-provider/resources/lambda/hello-world/build.gradle.ejs [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+dependencies {
+    compile (
+        'com.amazonaws:aws-lambda-java-core:1.2.0',
+        'com.amazonaws:aws-lambda-java-events:2.2.7'
+    )
+}
+
+jar.baseName = 'latest_build'
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/resources/lambda/hello-world/build.gradle.ejs]: If possible, we shouldn't be hardcoding this filename and instead get it from the package request. Maybe this SO post can help: https://stackoverflow.com/questions/38222724/rename-file-in-gradle-with-passed-parameter

File: packages/amplify-java-template-provider/resources/lambda/hello-world/pom.xml.ejs [hunk 18]
```diff
@@ -0,0 +1,46 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/resources/lambda/hello-world/pom.xml.ejs]: not needed now that gradle is used

File: packages/amplify-java-template-provider/src/providers/helloWorldProvider.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+      destMap: {
+        'hello-world/build.gradle.ejs': path.join('build.gradle'),
+        'hello-world/event.json': path.join('src', 'main', 'java', 'event.json'),
+        'hello-world/HelloPojo.java.ejs': path.join('src', 'main', 'java','example' ,'HelloPojo.java'),
+        'hello-world/RequestClass.java.ejs': path.join('src', 'main', 'java','example' , 'RequestClass.java'),
+        'hello-world/ResponseClass.java.ejs': path.join('src', 'main', 'java','example' , 'ResponseClass.java'),
+        'InvocationShim/build.gradle.ejs': path.join('src', 'Mock', 'build.gradle'),
+        'InvocationShim/MockContext.java.ejs': path.join('src', 'Mock', 'src', 'main', 'java','InvocationShim', 'MockContext.java'),
+        'InvocationShim/MockLogger.java.ejs': path.join('src', 'Mock', 'src', 'main', 'java','InvocationShim', 'MockLogger.java'),
+        'InvocationShim/Program.java.ejs': path.join('src', 'Mock', 'src', 'main', 'java','InvocationShim','Program.java'),
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/src/providers/helloWorldProvider.ts]: Again, if possible the shim should stay in the runtime provider. At the very least, it should not be in the src folder of the function directory.

File: packages/amplify-java-template-provider/src/providers/helloWorldProvider.ts [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    'InvocationShim/Program.java.ejs',
+  ];
+  return {
+    functionTemplate: {
+      sourceRoot: pathToTemplateFiles,
+      sourceFiles: files,
+      defaultEditorFile: path.join('src', 'main', 'java', 'HelloPojo.java'),
+      destMap: {
+        'hello-world/build.gradle.ejs': path.join('build.gradle'),
+        'hello-world/event.json': path.join('src', 'main', 'java', 'event.json'),
```
Comment (edwardfoyle) [inline packages/amplify-java-template-provider/src/providers/helloWorldProvider.ts]: the test event shouldn't be in the src folder

File: packages/amplify-provider-awscloudformation/lib/build-resources.js [hunk 21]
```diff
@@ -73,7 +73,7 @@ async function buildResource(context, resource) {
     packagePromise
       .then(result => {
         const packageHash = result.packageHash;
-        zipFilename = packageHash ? `${resource.resourceName}-${packageHash}-build.zip` : zipFilename;
+        zipFilename = packageHash ? `${resource.resourceName}-${packageHash}-build.zip` : result.zipFileName;
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/lib/build-resources.js]: undo this change

File: packages/amplify-java-runtime-provider/src/utils/invoke.ts [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+  const shimPath = path.join(request.srcRoot, 'src','Mock','build','libs');
+  const cp = childProcess.spawnSync('java', [ '-jar', 'Mock-all.jar' ,request.handler , request.event], { cwd: shimPath , stdio : 'pipe' , encoding: 'utf-8' });
+
+  let result  = JSON.stringify(cp.stdout);
+  try {
+    result = JSON.parse(result);
+  } catch (err) {
+    context.print.warning('Could not parse function output as JSON. Using raw output.');
+  }
+  return Promise.resolve({result});
```
Comment (edwardfoyle) [inline packages/amplify-java-runtime-provider/src/utils/invoke.ts]: Invoke should return just the return value of the underlying lambda function. returning a strigified version of stdout will include any log lines the lambda printed. Additionally, because it looks like your shim is printing `EVENT: <result>` it will include the EVENT prefix which we don't want. And once result is just the return value of the labmda, you should return `Promise.resolve(result)` (no braces) because it shouldn't be wrapped in another object

File: packages/amplify-java-function-runtime-provider/resources/localinvoke/build.gradle [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+         'com.amazonaws:aws-lambda-java-events:2.2.7',
+    )
+    compile fileTree(dir: 'lib', include: '*.jar')
+}
+
+jar {
+    manifest {
+        attributes 'Main-Class': 'InvocationShim.Program'
+    }
+}
```
Comment (edwardfoyle) [inline packages/amplify-java-function-runtime-provider/resources/localinvoke/build.gradle:22]: as discussed offline add
```
java {
    sourceCompatibility = JavaVersion.VERSION_11
    targetCompatibility = JavaVersion.VERSION_11
}
```

File: packages/amplify-java-function-runtime-provider/tsconfig.json [hunk 24]
```diff
@@ -0,0 +1,13 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "outDir": "./lib",
+    "rootDir": "./src",
+    "lib": [
```
Comment (edwardfoyle) [inline packages/amplify-java-function-runtime-provider/tsconfig.json:6]: I don't think this needs to be invluded

File: packages/amplify-java-function-template-provider/tsconfig.json [hunk 25]
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "outDir": "./lib",
+    "rootDir": "./src",
+    "strict": true,
```
Comment (edwardfoyle) [inline packages/amplify-java-function-template-provider/tsconfig.json:6]: this is in the root tsconfig

File: packages/amplify-provider-awscloudformation/lib/build-resources.js [hunk 26]
```diff
@@ -73,7 +73,7 @@ async function buildResource(context, resource) {
     packagePromise
       .then(result => {
         const packageHash = result.packageHash;
-        zipFilename = packageHash ? `${resource.resourceName}-${packageHash}-build.zip` : zipFilename;
+        zipFilename = packageHash ? `${resource.resourceName}-${packageHash}-build.zip` : zipFileName;
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/lib/build-resources.js:76]: this needs to be reverted

File: packages/amplify-java-function-runtime-provider/src/utils/detect.ts [hunk 27]
```diff
@@ -0,0 +1,54 @@
+import childProcess from 'child_process';
+import * as semver from 'semver';
+import { constants } from './constants';
+
+export async function checkJava() {
+  var javaSpawn = childProcess.spawnSync('java', ['-version'], {stdio : 'pipe' , encoding: 'utf-8' })
+  const regex = /(\d+\.)(\d+\.)(\d)/g;
+  if (javaSpawn.stderr !== null) {
+    let data : string = javaSpawn.output.toString().split(/\r?\n/)[0];
+    let javaVersion = data.match(regex);
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-java-function-runtime-provider/src/utils/detect.ts:15]: If you have time to update this to use semver that would be great. Otherwise it can wait until later
